File: Graph/flightDiscount/flightDiscount.py
# 3 4
# 1 2 3
# 2 3 1
# 1 3 7
# 2 1 5

import heapq

if __name__ == "__main__":
    inp1 = input().split(' ')
    n, e = int(inp1[0]), int(inp1[1])

    graph = dict()
    for i in range(n): graph[i+1] = []

    for _ in range(e):
        inp2 = input().split(' ')
        e1, e2, w = int(inp2[0]), int(inp2[1]), int(inp2[2])
        graph[e1].append((w, e2))

    q = []
    dist = [[float('inf') for _ in range(n+1)] for _ in range(2)]
    heapq.heappush(q, (0, 1, 0))

    while len(q) > 0:
        d, element, used = heapq.heappop(q)
        if d > dist[used][element]: continue

        for w, neighbor in graph[element]:
            if d + w < dist[used][neighbor]:
                dist[used][neighbor] = d + w
                heapq.heappush(q, (d + w, neighbor, used))
            if used == 0 and d + (w // 2) < dist[1][neighbor]:
                dist[1][neighbor] = d + (w // 2)
                heapq.heappush(q, (d + (w // 2), neighbor, 1))

    print(min(dist[0][n], dist[1][n]))

File: Graph/flightDiscount/flightDiscount.cpp
#include <iostream>
#include <vector>
#include <array>
#include <queue>
#include <tuple>
#include <climits>
#include <algorithm>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    // adjacency: for each node, list of (neighbor, weight)
    vector<vector<pair<int, int>>> graph(n + 1);
    for (int i = 0; i < m; i++) {
        int a, b, w;
        cin >> a >> b >> w;
        graph[a].emplace_back(b, w);
    }

    const long long INF = LLONG_MAX;
    // d[node][used], used = 0 (no coupon yet) / 1 (coupon spent)
    vector<array<long long, 2>> d(n + 1, {INF, INF});
    d[1][0] = 0;

    // min-heap of (distance, node, used)
    priority_queue<tuple<long long, int, int>,
                   vector<tuple<long long, int, int>>,
                   greater<>> pq;
    pq.emplace(0LL, 1, 0);

    while (!pq.empty()) {
        auto [dc, u, used] = pq.top();
        pq.pop();
        if (dc > d[u][used]) continue;

        for (auto [v, w] : graph[u]) {
            // take edge at full price
            if (dc + w < d[v][used]) {
                d[v][used] = dc + w;
                pq.emplace(d[v][used], v, used);
            }
            // use the coupon on this edge (only if not used yet)
            if (used == 0) {
                long long nd = dc + w / 2;
                if (nd < d[v][1]) {
                    d[v][1] = nd;
                    pq.emplace(nd, v, 1);
                }
            }
        }
    }

    cout << min(d[n][0], d[n][1]) << "\n";
    return 0;
}
